File: src/main/java/net/guizhanss/fastmachines/libs/guizhanlib/localization/Localization.java
package net.guizhanss.fastmachines.libs.guizhanlib.localization;

import java.io.File;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;

import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

import net.guizhanss.fastmachines.libs.guizhanlib.utils.ChatUtil;

/**
 * A multi-language, YAML-file-backed localization service. Colors every returned string/list via
 * {@link ChatUtil}. Languages are checked in the order they were added via {@link #addLanguage};
 * the first one containing a non-empty value for a path wins.
 *
 * Java-8-safe port of GuizhanLib's {@code Localization}/{@code MinecraftLocalization}.
 */
public class Localization {

    protected final JavaPlugin plugin;
    private final String folderName;
    private final List<String> languages = new ArrayList<>();
    private final Map<String, Language> langMap = new HashMap<>();
    private final File langFolder;

    public Localization(JavaPlugin plugin) {
        this(plugin, "lang");
    }

    public Localization(JavaPlugin plugin, String folderName) {
        this.plugin = plugin;
        this.folderName = folderName;
        this.langFolder = new File(plugin.getDataFolder(), folderName);
        if (!plugin.getDataFolder().exists()) {
            plugin.getDataFolder().mkdirs();
        }
        if (!langFolder.exists()) {
            langFolder.mkdirs();
        }
    }

    public void addLanguage(String langFilename) {
        File langFile = new File(langFolder, langFilename + ".yml");
        String resourcePath = folderName + "/" + langFilename + ".yml";

        if (!langFile.exists()) {
            try {
                plugin.saveResource(resourcePath, false);
            } catch (IllegalArgumentException ex) {
                plugin.getLogger().log(Level.SEVERE, "The default language file {0} does not exist in jar file!", resourcePath);
                return;
            }
        }

        languages.add(langFilename);

        InputStream resource = plugin.getResource(resourcePath);
        if (resource == null) {
            return;
        }
        FileConfiguration defaultConfig;
        try (InputStreamReader reader = new InputStreamReader(resource, StandardCharsets.UTF_8)) {
            defaultConfig = YamlConfiguration.loadConfiguration(reader);
        } catch (java.io.IOException e) {
            return;
        }
        langMap.put(langFilename, new Language(langFilename, langFile, defaultConfig));
    }

    public String getString(String path) {
        for (String lang : languages) {
            Language language = langMap.get(lang);
            if (language == null) {
                continue;
            }
            String value = language.getConfig().getString(path);
            if (value != null) {
                return ChatUtil.color(value);
            }
        }
        return "";
    }

    public List<String> getStringList(String path) {
        for (String lang : languages) {
            Language language = langMap.get(lang);
            if (language == null) {
                continue;
            }
            List<String> value = language.getConfig().getStringList(path);
            if (value != null && !value.isEmpty()) {
                return ChatUtil.color(value);
            }
        }
        return new ArrayList<>();
    }
}

File: src/main/kotlin/net/guizhanss/fastmachines/libs/guizhanlib/localization/Localization.kt
package net.guizhanss.fastmachines.libs.guizhanlib.localization

import net.guizhanss.fastmachines.libs.guizhanlib.utils.ChatUtil
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.logging.Level

/**
 * A multi-language, YAML-file-backed localization service. Colors every returned string/list via
 * [ChatUtil]. Languages are checked in the order they were added via [addLanguage]; the first one
 * containing a non-empty value for a path wins.
 *
 * Java-8-safe, Kotlin port of GuizhanLib's `Localization`/`MinecraftLocalization` - see the
 * package-level note in [net.guizhanss.fastmachines.libs.guizhanlib].
 */
open class Localization(
    protected val plugin: JavaPlugin,
    private val folderName: String = "lang",
) {

    private val languages = mutableListOf<String>()
    private val langMap = mutableMapOf<String, Language>()
    private val langFolder = File(plugin.dataFolder, folderName)

    init {
        if (!plugin.dataFolder.exists()) {
            plugin.dataFolder.mkdirs()
        }
        if (!langFolder.exists()) {
            langFolder.mkdirs()
        }
    }

    fun addLanguage(langFilename: String) {
        val langFile = File(langFolder, "$langFilename.yml")
        val resourcePath = "$folderName/$langFilename.yml"

        if (!langFile.exists()) {
            try {
                plugin.saveResource(resourcePath, false)
            } catch (ex: IllegalArgumentException) {
                plugin.logger.log(Level.SEVERE, "The default language file {0} does not exist in jar file!", resourcePath)
                return
            }
        }

        languages.add(langFilename)

        val resource = plugin.getResource(resourcePath) ?: return
        val defaultConfig = InputStreamReader(resource, StandardCharsets.UTF_8).use {
            YamlConfiguration.loadConfiguration(it)
        }
        langMap[langFilename] = Language(langFilename, langFile, defaultConfig)
    }

    open fun getString(path: String): String {
        for (lang in languages) {
            val value = langMap[lang]?.config?.getString(path)
            if (value != null) return ChatUtil.color(value)
        }
        return ""
    }

    open fun getStringList(path: String): List<String> {
        for (lang in languages) {
            val value = langMap[lang]?.config?.getStringList(path)
            if (!value.isNullOrEmpty()) return ChatUtil.color(value)
        }
        return emptyList()
    }
}
